### ISSUE TITLE
Cannot handle special characters (em dash, emoji, etc.) 

### ISSUE THREAD

**Issue by neurodynamic** [jxxcarlson/elm-tar#3]
On trying to extract `.tar` files that contain these kinds of characters and were created with `elm-tar`, I get a `(Error 22 - Invalid argument.)` error message, or a message about the `.tar` file being corrupt, depending on the extraction program used.

**Comment by malaire** [comment]
There is an error in handling file size.

In [Tar.elm#L705](https://github.com/jxxcarlson/elm-tar/blob/3.0.9/src/Tar.elm#L705) `metadata.fileSize` is set to `String.length contents`, which returns size in UTF-16 code units, not in bytes. Then in [Tar.elm#L775](https://github.com/jxxcarlson/elm-tar/blob/3.0.9/src/Tar.elm#L775) that `fileSize` is placed into tar header which requires file size in bytes, so header will have incorrect file size if file contains any character which takes more than one byte when encoded in UTF-8 - i.e. only ASCII will work correctly.

(There might be other errors also.)

**Comment by malaire** [comment]
I think that `encodeTextFile` should be changed to just:
```
encodeTextFile : MetaData -> String -> Encode.Encoder
encodeTextFile metaData_ contents =
    encodeBinaryFile metaData_ (Encode.string contents)
```

**Comment by malaire** [comment]
There is also another error - `normalizeString` at [Tar.elm#L986](https://github.com/jxxcarlson/elm-tar/blob/3.0.9/src/Tar.elm#L986) normalizes to amount of UTF-16 code units, but it is used in many places where normalization must be in bytes.

**Comment by malaire** [comment]
`padContents` at [Tar.elm#L893](https://github.com/jxxcarlson/elm-tar/blob/3.0.9/src/Tar.elm#L893) is also wrong, but it is used only in `encodeTextFile`, so if `encodeTextFile` is changed as I suggested above, `padContents` can be just removed.

**Comment by folkertdev** [comment]
I implemented the first fix you suggested @malaire. The second problem is a little trickier, because string functions implicitly use `String.length`. Any thoughts on a byte-based `normalizeString`?

**Comment by folkertdev** [comment]
I now have 

```elm
normalizeString : Int -> String -> String
normalizeString desiredLength str =
    case desiredLength of
        0 ->
            -- just to be safe. otherwise unbounded recursion
            ""

        _ ->
            let
                dropped =
                    str
                        -- first `desiredLength` characters
                        -- but this function should produce
                        -- `desiredLength` bytes, not characters
                        |> String.left desiredLength
                        -- so this step is required
                        |> dropLeftLoop (desiredLength - 1)

                paddingSize =
                    desiredLength - Encode.getStringWidth dropped
            in
            dropped ++ String.repeat paddingSize "\u{0000}"


dropLeftLoop : Int -> String -> String
dropLeftLoop desiredLength str =
    if Encode.getStringWidth str > desiredLength then
        dropLeftLoop desiredLength (String.dropRight 1 str)

    else
        str
```
Strings must be null-terminated, so 32 bytes actually only allows 31 ascii characters, then a null byte.
That's why `dropLeftLoop (desiredLength - 1)`. 

The implementation conforms to these tests:

```elm
            [ fuzz (Fuzz.tuple ( Fuzz.intRange 0 300, Fuzz.string )) "string is no longer than expected" <|
                \( size, string ) ->
                    normalizeString size string
                        |> Encode.getStringWidth
                        |> Expect.equal size
            , fuzz (Fuzz.tuple ( Fuzz.intRange 1 300, Fuzz.string )) "string ends with null" <|
                \( size, string ) ->
                    normalizeString size string
                        |> String.endsWith "\u{0000}"
                        |> Expect.equal True

```

**Comment by malaire** [comment]
I think there is one potential problem there. String does not contain characters, but UTF-16 code units. Some characters (e.g. many emoji) take two UTF-16 code units, and if you split the string in between those two code units, result will be invalid Unicode.

**Comment by malaire** [comment]
I think the solution is to check last item of `dropped` - if it is between 0xDC00 - 0xDFFF (inclusive), that needs to be dropped also.

**Comment by malaire** [comment]
In UTF-16 "surrogate pair", first code point is between 0xD800 - 0xDBFF, and second code point is between 0xDC00 - 0xDFFF. So String must never end with code point between 0xD800 - 0xDBFF.

https://en.wikipedia.org/wiki/UTF-16

**Comment by malaire** [comment]
But `String.dropRight` claims to drop _characters_ - so I'm not sure how Elm actually handles the case of surrogate pairs. https://package.elm-lang.org/packages/elm/core/latest/String#dropRight

If `String.dropRight 1 str` does the right thing, and drops one character, i.e. one or two UTF-16 code units as needed, then there is nothing more to fix.

**Comment by jxxcarlson** [comment]
I’ll look into it — thanks


Jim
https://jxxcarlson.io

**Comment by jxxcarlson** [comment]
Merged Folkert's fix -- hopefully ok now.

**Comment by jxxcarlson** [comment]
I'll leave this open for a few days in case any further problems show up.

**Comment by malaire** [comment]
Unfortunately `String.dropRight` does NOT do the right thing.
See https://github.com/elm/core/issues/1061 and https://github.com/elm/core/issues/977

**Comment by folkertdev** [comment]
Could we use [BrianHicks/elm-string-graphemes](https://package.elm-lang.org/packages/BrianHicks/elm-string-graphemes/latest/String-Graphemes)? It claims to drop actual characters.

**Comment by malaire** [comment]
`BrianHicks/elm-string-graphemes` goes beoynd Unicode characters and handles Unicode graphemes (which is often what average person might consider as a "character"). So yes, that would be really good solution, as long as performance is good enough.

For an example of difference between Unicode character and grapheme, "à" can be encoded in Unicode in two different ways. Either as a single Unicode character/grapheme "à" or as a combination of two Unicode characters "a" and "`", which together form a single grapheme.

ps. Unicode is really complex thing to handle.

**Comment by folkertdev** [comment]
I think performance should be fine. This is just for the metadata. I'll have a got at using that library. 

Could you maybe give some concrete examples (to be turned into test cases) where the current implementation would fail? I assume there is some `foo\u{xxxx}\u{xxxx}` where if we want at most 4 bytes used, it should give back `foo` but currently would give `foo\u{xxxx}`. I'd like some concrete examples of what `xxxx` would look like

**Comment by malaire** [comment]
For example string "ab\u{1F9B8}\u{1F3FD}cd" padded to 9 bytes as UTF-8.

When handling Unicode characters but not graphemes correct result is:
[0x61, 0x62, 0xf0, 0x9f, 0xa6, 0xb8, 0x00, 0x00, 0x00]

When handling graphemes correct result is:
[0x61, 0x62, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]

This string has superhero-emoji followed by a modifier which changes skin color of the emoji. When not handling graphemes, the skin-color-modifier can be dropped without the emoji, but when handling graphemes both must be dropped as a unit.

I think current implementation results in following **invalid** value, because it only handles UTF-16 code units, which means it outputs emoji correctly, but then drops one UTF-16 code unit of skin-color-modifier which consists of two UTF-16 code units:
[0x61, 0x62, 0xf0, 0x9f, 0xa6, 0xb8, 0xed, 0xa0, 0xbc]

**Comment by malaire** [comment]
Since main problem here is shortening the string, one option would be to only accept strings which fit, and return an error if string is too long.

**Comment by folkertdev** [comment]
@neurodynamic is your problem solved by elm-tar 4.0.0?

**Comment by neurodynamic** [comment]
@folkertdev Yup, seems to be all working. Thanks for all the help; I would have had a horrible time figuring this out on my own!
